### PR TITLE
Here, fixes locking properly.

### PR DESCRIPTION
**Pull request by PJB3005**
closes #13065 

~~Not yet tested my DD server is literally starting as I type this.~~

### CODE REVIEW

**Comment by sonime** [comment]
So this is 💯?

**Review by ghost — CHANGES_REQUESTED** [review]
for no reason

File: code/game/atoms_movable.dm [hunk 1]
```diff
@@ -299,19 +300,56 @@
...
+
 	if (locking_categories_name.Find(category))
 		var/datum/locking_category/C = locking_categories_name[category]
 		return C.locked
 
 	return list()
 
-/atom/movable/proc/is_locking(var/category) // Returns true if we have any locked atoms in this category.
+// Returns true if we have any locked atoms in this category.
+// If this returns TRUE, it means that get_locked with the same parameters will return a list with at least a size of 1.
```
Comment (9600bauds) [inline code/game/atoms_movable.dm]: Why this return inconsistency?

File: code/game/objects/structures/vehicles/wheelchair.dm [hunk 2]
```diff
@@ -169,7 +169,8 @@
 /obj/structure/bed/chair/vehicle/wheelchair/multi_people/examine(mob/user)
 	..()
 
-	if(locked_atoms.len > 9)
+	var/list/L = get_locked(is_locking(/datum/locking_category/buckle/chair/vehicle))
```
Comment (ComicIronic) [inline code/game/objects/structures/vehicles/wheelchair.dm]: Did you mean to do this? It doesn't look like it would work.

Comment (9600bauds) [reply code/game/objects/structures/vehicles/wheelchair.dm]: I think it's supposed to be the weird return I pointed out above

File: code/modules/hydroponics/spreading/spreading_growth.dm [hunk 3]
```diff
@@ -63,8 +63,8 @@
 	update_icon()
 
 	if(is_mature() && special_cooldown())
-		if(locked_atoms && locked_atoms.len)
-			var/mob/V = locked_atoms[1]
+		if(is_locking(/datum/locking_category/plantsegment))
+			var/mob/V = get_locked(/datum/locking_category/plantsegment)[1]
```
Comment (9600bauds) [inline code/modules/hydroponics/spreading/spreading_growth.dm]: Should be locate in get_locked maybe

File: code/game/objects/structures/cage.dm [hunk 4]
```diff
@@ -223,7 +223,11 @@
 			to_chat(victim, "<span class='notice'>You suddenly find yourself locked in a cage!</span>")
 
 /obj/structure/cage/proc/mob_is_inside(mob/checked)
-	return ((contents.Find(checked)) || (locked_atoms.Find(checked)))
+	if (contents.Find(checked))
+		return TRUE
+
+	var/list/locked = get_locked(/datum/locking_category/cage)
+	return locked.Find(checked)
```
Comment (9600bauds) [inline code/game/objects/structures/cage.dm]: Weird that this isn't a helper proc in itself

File: code/modules/hydroponics/spreading/spreading_growth.dm (hunk 3 above)
Comment (PJB3005) [reply code/modules/hydroponics/spreading/spreading_growth.dm]: Doesn't really matter.

File: code/game/objects/structures/vehicles/wheelchair.dm (hunk 2 above)
Comment (PJB3005) [reply code/game/objects/structures/vehicles/wheelchair.dm]: uh, fuck.

Hey I wasn't done testing it yet.

File: code/game/objects/structures/cage.dm (hunk 4 above)
Comment (PJB3005) [reply code/game/objects/structures/cage.dm]: Actually I can write this more cleanly.

File: code/modules/hydroponics/spreading/spreading_growth.dm (hunk 3 above)
Comment (9600bauds) [reply code/modules/hydroponics/spreading/spreading_growth.dm]: Only if nothing in the future ever happens to make it matter

what a gamble!

Comment (PJB3005) [reply code/modules/hydroponics/spreading/spreading_growth.dm]: Actually you're right I can think of where this could be useful.

**Comment by Probe1** [comment]
seems fine 🤷‍♂️

**Comment by ghost** [comment]
This is absolutely worse in functionality than my PR for several reasons.

**Comment by Probe1** [comment]
If you two don't start discussing this _in complete sentences that describe what you are doing and why_ I'm going with PJBs and we're moving forward to fixing the snow map.

**Comment by ghost** [comment]
There are some pros, which definitely should be added to the game
- get locking is better, but that won't conflict with what I have

But in terms of optimisation, you're missing some of the crucial things

- Lists don't get deleted after being unlocked 
- Lists get intialised when they don't need to be
- 

I don't get why you called my fix half assed when this is worse though

Sorry for not making this post instantly, I was typing up my reasoning

File: code/game/atoms_movable.dm [hunk 5]
```diff
@@ -262,6 +260,7 @@
 	return 1
 
 /atom/movable/proc/unlock_atom(var/atom/movable/AM)
+	locking_init()
```
Comment (ghost) [inline code/game/atoms_movable.dm]: Why are you initialising three lists here? You really shouldn't be. You're UNLOCKING an atom, this is your chance to get rid of unused lists, not create them.

File: code/game/atoms_movable.dm [hunk 6]
```diff
@@ -289,6 +289,7 @@
 	locking_categories += C
 
 /atom/movable/proc/get_lock_cat(var/category = /datum/locking_category)
+	locking_init()
```
Comment (ghost) [inline code/game/atoms_movable.dm:291]: You shouldn't intialised a list here either. If it's just initialised, it's always going to return null

File: code/game/atoms_movable.dm [hunk 7]
```diff
@@ -299,19 +300,51 @@
...
-/atom/movable/proc/get_locked(var/category)
+// Returns the locking category for a locked atom.
+// Returns null if the object is not locked to this.
+/atom/movable/proc/get_lock_cat_for(var/atom/movable/AM)
+	locking_init()
+	return locked_atoms[AM]
+
+// Returns a list (yes, always a list!) of things locked to this category.
+/atom/movable/proc/get_locked(var/category, var/subtypes = FALSE)
+	locking_init()
```
Comment (ghost) [inline code/game/atoms_movable.dm]: Again, there should be no init here.

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by 9600bauds** [comment]
I was going to comment on keeping empty lists around but figured it wasn't all that much of a big deal probably, where are the lists needlessly initialized though?

**Comment by ghost** [comment]
@9600bauds the whole point is to get rid of unnecessary lists, but this PR only goes half way to achieving that

**Comment by PJB3005** [comment]
@DrCelt you are retarded. Fuck off.

**Comment by 9600bauds** [comment]
and here we go again

**Comment by PJB3005** [comment]
I can optimize the access of `get_locked` and such, but otherwise Celt's a damn moron.

**Comment by PJB3005** [comment]
The point of lazy initialization is to be more efficient on resources. You won't create something that won't be used. Not having a million fucking lists be spawned is good because 99.999% of those wouldn't be used anyways.

If you honestly think that there's a need to delete lists once they run empty you're retarded. Mark my fucking word with this fix in we'll never run out of lists anymore before we have a fucking remake, unless you do something really stupid.

So stop trying to be so smart and calling my fix "inferior" and stop talking about things you clearly don't understand.

And fuck off too.

**Comment by Intigracy** [comment]
Baddy Paddy versus Belgium, regardless of who wins, we all lose.

**Comment by 9600bauds** [comment]
What, so we HAVE to skimp on single vars and insignificantly unprofilable syntax, but keeping all these empty lists around is A-OK because "lol we'll probably never have a problem until the remake"

You said you'd do it, go ahead and fuckin' do it. Celt's concerns are entirely legitimate and you know it'd take minutes to fix.

**Comment by PJB3005** [comment]
> legitimate

Nice fucking meme.

**Review by 9600bauds — DISMISSED** [review]
Said he would do it properly, needs to just do it properly so we can get this over with.

**Comment by 9600bauds** [comment]
Listen, getting you to write a sensible post was a childbirth. You already devolved back into autistic screeching and I won't go through that effort again.

So just fucking fix it. You said you would.

**Comment by PJB3005** [comment]
I did fucking do it, though?

Celt gets pissy over what's probably less than a hundred lists or so, which is fucking nothing. Me "properly doing it" was never intended to be as holy memory efficient at our shitposter here.

**Comment by 9600bauds** [comment]
Yes. I don't care what's up between you and Celt. You've ragged people out about stuff even more insignificant before and this time it's actually trivial to fix (and you know how to). No lazy code.

**Comment by PJB3005** [comment]
It's also irrelevant to fix.

And if I knew celt isn't doing this because he needs to be a shitter, I maybe would've.

**Comment by 9600bauds** [comment]
I just told you I don't care what you think of Celt and I just told you to fix it.

**Comment by 9600bauds** [comment]
Hell you even said above

>I can optimize the access of get_locked

**Comment by PJB3005** [comment]
Yes, and I will do that.

**Comment by PJB3005** [comment]
The other part celt said is fucking retarded though.

**Comment by ghost** [comment]
Why have a hundred or so lists when you can have none

**Comment by PJB3005** [comment]
@DrCelt because fuck you.

Unused RAM is wasted RAM.

**Comment by 9600bauds** [comment]
aahah what the fuck?

That's Dylan-"copypaste makes the code clearer" tier

holy fuck I'm laughing now

File: code/game/atoms_movable.dm (hunk 5 above)
Comment (9600bauds) [reply code/game/atoms_movable.dm]: Correct, the three lists should be nulled here

File: code/game/atoms_movable.dm (hunk 6 above)
Comment (9600bauds) [reply code/game/atoms_movable.dm:291]: Correct, if locking is not initialized just return null or FALSE, this is wasteful 4nr and would probably have taken the same amount of time to do it properly than to paste this here

File: code/game/atoms_movable.dm (hunk 7 above)
Comment (9600bauds) [reply code/game/atoms_movable.dm]: Correct, if locking is not initialized just return an empty list, this is wasteful 4nr and would probably have taken the same amount of time to do it properly than to paste this here

**Review by 9600bauds — CHANGES_REQUESTED** [review]
very very easy fixes for very very lazy coding, can we please just get all of this over with nice and easy and fast

**Comment by PJB3005** [comment]
If you're gonna be Petty over the locking lists, I can be just as Petty over the CPU usage from needing to recreate lists from repeated locks.

**Comment by 9600bauds** [comment]
We get enough of you being Petty daily, you could have fixed this ten times already, let's not waste everyone's time anymore.

And besides, that pettiness wouldn't even be correct, because there's still no excuse for initializing the lists in get_locked(), which will happen very often on things that didn't have anything locked.

**Comment by PJB3005** [comment]
The incorrect init is something I've already said I was gonna fix no matter what.

**Comment by TheFriggenSawbones** [comment]
pjb has a point, rAM is a good album. While I like Discovery better lets put this into space
https://www.youtube.com/watch?v=XfH3erWacsQ

**Comment by ComicIronic** [comment]
Lazy loading takes less CPU for creation because you only have lists as needed. It's generally better performance in all areas.

**Comment by ghost** [comment]
@PJB3005 who's being pretty about the locking lists?

**Comment by PJB3005** [comment]
@DrCelt you.

**Comment by ghost** [comment]
... How?

**Comment by ghost** [comment]
@PJB3005 Chill out dude and take requests to the credit.

**Comment by 9600bauds** [comment]
👌 

c comment in get_lock_cat() though

**Review by 9600bauds — APPROVED** [review]

File: code/game/atoms_movable.dm (hunk 6 above)
Comment (9600bauds) [reply code/game/atoms_movable.dm:291]: What happened here?

**Comment by ghost** [comment]
There's still some optimisations to be made here but that doesn't mean this shouldn't be merged. I will change #13065 to include those instead of my current changes

**Comment by 9600bauds** [comment]
Well this is still not nulling the lists AND still initializing them in get_lock_cat() so honestly what the fuck.

**Comment by PJB3005** [comment]
@9600bauds if you were to read the code, you would see that `get_lock_cat` initializes the category if it doesn't exist yet.

So yes, it's needed.

**Comment by sonime** [comment]
Bunch of fucking babies that makes Congress look mature.

**Comment by Probe1** [comment]
Please stop provoking them.

**Comment by sonime** [comment]
They're doing a good job of that themselves.

**Comment by ComicIronic** [comment]
Oh my god it's TermedClepe 2.0

**Comment by sonime** [comment]
?

**Comment by Kitfox88** [comment]
lol

**Comment by d3athrow** [comment]
bork

**Comment by sonime** [comment]
@PJB3005 http://puu.sh/t1YYY/ba40a17d39.jpg

**Comment by sonime** [comment]
@PJB3005 Holy word of Pomf

**Comment by 9600bauds** [comment]
https://www.youtube.com/watch?v=LP0HYIkHs2Q

**Comment by Senaat1** [comment]
DEDITATED
WAAAM

**Comment by ghost** [comment]
Hi there @PJB3005 - please edit the PR's description from closes #13065, that PR has content in it other than what you're replacing, thanks

**Comment by 9600bauds** [comment]
Celt, like 95% of that PR would conflict with this one anyways (and 95% of the comments would be outdated/best forgotten too). It'd be best if you just opened a new one afterwards.

**Comment by clusterfack** [comment]
Lists aren't the reason we're hitting the ram limit anyway, it's images.

Doesn't the snow map have some sort of shadow image on everything?

**Comment by 9600bauds** [comment]
I thought we had ran out of ram due to lists back when we went 512x512 for the first time (space turfs had locking lists), and the problem was reappearing because snow turfs had them again

How wrong am I

**Comment by sonime** [comment]
@clusterfack The shadows are done with matrices IIRC.

**Comment by ghost** [comment]
@clusterfack We ran out of lists; I read the runtimes. We've ran out on lists on the live server before, This is the reason why.
I don't believe shadows has anything to do with it.

**Comment by Probe1** [comment]
Just add what you feel it doesn't address on top of it @DrCelt

**Comment by clusterfack** [comment]
@sonime 
If you can see it ingame, it has an image. From the memory stats pomf has shown me in recent times, those are the causes of crashes from running out of memory.

@9600bauds 
Only atom movables have locking lists, so completely off the mark, and I cant even imagine a world where snow turfs have locking lists

@DrCelt 
If you're talking about a runtime solely about running out of lists, which I would be surprised if such a thing existed, then yes this would help alleviate that. But we've had a LOT more lists than this in the past at several points, since atom movables make up about 250,000 instances, and atoms themselves make up about 2 million and we've had lists on atoms in the past without issue.

If there is such an issue one of you guys probably dun fucked it up by adding some self replicating list somewhere and this would be a bandaid to that.

**Comment by ghost** [comment]
@clusterfack Unfortunately on this issue all I can say is that before a fix similar to this it runtimed due to lack of lists and after it, it didn't. Another fix in the other PR was that of mineral turfs having two lists on them which doesn't help matters

**Comment by clusterfack** [comment]
@DrCelt Which means you keep bandaiding it without finding the real problem, thats literally all you're saying to me.

**Pull request merged**